src/pf/optimize.lua

```diff
@@ -176,15 +176,24 @@ local function simplify(expr, is_tail)
       if type(lhs) == 'number' and type(rhs) == 'number' then
          return assert(folders[op])(lhs, rhs)
       elseif associative_binops[op] then
-         if type(rhs) == 'table' and rhs[1] == op and type(lhs) == 'number' then
+         -- Try to make the right operand a number.
+         if type(lhs) == 'number' then
```

Comment:
Let's only do this if the RHS isn't a number.  Otherwise we introduce arbitrary mutations to the graph depending on how many times the thing is run.